src/Core/Coverage.idr

```diff
@@ -148,7 +148,7 @@ isEmpty defs env _ = pure False
 altMatch : CaseAlt vars -> CaseAlt vars -> Bool
 altMatch _ (DefaultCase _) = True
 altMatch (DelayCase _ _ t) (DelayCase _ _ t') = True
-altMatch (ConCase _ t _ _) (ConCase _ t' _ _) = t == t'
+altMatch (ConCase n t _ _) (ConCase n' t' _ _) = t == t' && n == n'
```

Comment:
I'd be happy for it to be merged but I am still in two minds about this change in the coverage checker.

It is clearly correct and fixes an existing bug but I was under the impression that coverage checking implicitly assumes that all patterns passed to it are already well-typed. So I'd prefer if ill-typed patterns like

```
Bug : Not (3 `LessThanOrEqualTo` 5)
Bug Refl impossible
```

are rejected before the coverage checker is called. On the other hand, that would require a much more involved change and it may be a while before somebody finds the time to work on it.

Reply:
Sorry, I chased this stuff back and saw the `mkTerm` issue, but I didn't see what else was happening in `Impossible.idr` aside from `mkTerm` itself.  It seems that most of that checking is already in place there, but line 48 is short circuiting it:

https://github.com/idris-lang/Idris2/blob/f840d1b771be59000b478780dcfc1dfa2bfde624/src/TTImp/Impossible.idr#L45-L52

I'm not sure why the type check is being short circuited, but if I remove that line and remove the change you quoted above, all of the tests pass. 

Further, for the `IAlternative` case, I didn't realize that I may have the type at hand (as a `Maybe (Closure [])`). I think I can take the same approach instead of picking the default alternative. (The type is indeed there in my test cases.)

A third thing - I tracked down the weird covering error that I was hitting when trying to fix frex (it was covering if I put ` = ?rhs` but not covering if I put `impossible`).  It appears that line 1131 of `ProcessDef.idr` silently drops an impossible clause when an error is raised in `getImpossibleTerm`.  In the case of frex, it raises an `AmbiguousName` because of `Ultimate`:

https://github.com/idris-lang/Idris2/blob/f840d1b771be59000b478780dcfc1dfa2bfde624/src/TTImp/ProcessDef.idr#L1126-L1133

I think we would want to give an error or warning if an impossible clause is discarded due to an error.

Reply:
I've reverted the changes to constructor checking in `Coverage.idr` and instead changed `dropNoMatch` to also check cases where there is only one definition matching the name.  It uses `TTImp.Impossible.match` to see if constructors are valid. Previously this was only used in the case of an ambiguous name to guess the correct name.

Note that because we're dealing with raw terms, it is not precise type checking. It looks like it chases down the type constructor in the return type and checks if it has the same name as the expected type.

As I noted above, the impossible clause is silently dropped if it doesn't match. That is the existing behavior for cases where the name many or zero options with an acceptable type. For example, in the current Idris the following line in frex is silently dropped because there are two `Refl` in scope:
```idris
  sumDegenerate a (x :: xs) (FS i) prf | Left  Refl impossible
```
It would be nice if this were a warning or error, but perhaps that should be a separate PR.

Reply:
I think I've resolved mjustus' concerns, but I don't know who is supposed to mark resolve conversation.

For IAlternative, I have code that checks each option, but it risks being exponential (unlikely), so I've left that at choosing the default.